### PR TITLE
Create ocsvm.jl

### PR DESCRIPTION
**Pull request by TheMrGhostman**
creating ocsvm ensemble for image data.

### CODE REVIEW

File: scripts/experiments_images/ocsvm.jl [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		default = 10
+		help = "number of anomaly classes"
+end
+parsed_args = parse_args(ARGS, s)
+@unpack dataset, max_seed, anomaly_classes = parsed_args
+
+modelname = "OCSVM"
+
+function sample_params()
+	par_vec = (round.([10^x for x in -4:0.1:2],digits=5),["poly", "rbf", "sigmoid"])
```
Comment (vitskvara) [inline scripts/experiments_images/ocsvm.jl]: Add the `nu` parameter with values `[0.01, 0.5, 0.99]`

File: scripts/experiments_images/ocsvm.jl [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    return Flux.mean(joint_pred, dims=1)
+end
+
+function fit(data, parameters, n_models=10)
+    # construct model - constructor should only accept kwargs
+    models = [GenerativeAD.Models.OCSVM(;parameters...) for _ = 1:n_models]
+	
+	# sumbsample and fit train data
+	tr_data = data[1][1]
+    M,N = size(tr_data)
```
Comment (vitskvara) [inline scripts/experiments_images/ocsvm.jl]: This is not gonna work, image data have size = (w,h,c,n).

File: scripts/experiments_images/ocsvm.jl [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    for (i,model) in enumerate(models)
+        joint_pred[i,:] = predict(model, data)
+    end
+    return Flux.mean(joint_pred, dims=1)
+end
+
+function fit(data, parameters, n_models=10)
+    # construct model - constructor should only accept kwargs
+    models = [GenerativeAD.Models.OCSVM(;parameters...) for _ = 1:n_models]
+
```
Comment (vitskvara) [inline scripts/experiments_images/ocsvm.jl:56]: What I would do is that I would reshape all data here, e.g. `data[1][1] = reshape(data[1][1], :, size(data[1][1],4))` (the same for (val and tst data).

Comment (vitskvara) [reply scripts/experiments_images/ocsvm.jl:56]: Then the rest is gonna work as you have it written already.

File: scripts/experiments_images/ocsvm.jl [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+while try_counter < max_tries
+	parameters = sample_params()
+
+	for seed in 1:max_seed
+		for i in 1:anomaly_classes
+			savepath = datadir("experiments/images/$(modelname)/$(dataset)/ac=$(i)/seed=$(seed)")
+
+			data = GenerativeAD.load_data(dataset, seed=seed, anomaly_class_ind=i)
+
```
Comment (vitskvara) [inline scripts/experiments_images/ocsvm.jl]: Ignore my second comment, it is actually better to reshape the data here, e.g. write a nice function for it.

File: scripts/experiments_images/ocsvm.jl [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	end
+
+	# construct return information - put e.g. the model structure here for generative models
+	training_info = (
+		fit_t = fit_t,
+		model = nothing
+		)
+
+	# now return the different scoring functions
+	training_info, [(x->joint_predictiion(models, x), parameters)]
```
Comment (vitskvara) [inline scripts/experiments_images/ocsvm.jl]: typo

File: scripts/experiments_images/ocsvm.jl [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    info = []
+    for (model, data) in zip(models, data_splits)
+        push!(info, fit!(model, data))
+    end
+    return info
+end
+
+function joint_prediction(models, data)
+    joint_pred = Array{Float32}(undef, length(models), size(data,2))
+    for (i,model) in enumerate(models)
```
Comment (vitskvara) [inline scripts/experiments_images/ocsvm.jl]: Is it not more economical to do the prediction here sample-wise? That way you only need to allocate a vector as you can already do the average in each step.

**Review by vitskvara — CHANGES_REQUESTED** [review]
Also add `ocsvm_run.sh` file

File: scripts/experiments_images/ocsvm.jl (hunk 2 above)
Comment (TheMrGhostman) [reply scripts/experiments_images/ocsvm.jl:59]: it will work because i reshaped images before

File: scripts/experiments_images/ocsvm.jl (hunk 5 above)
Comment (TheMrGhostman) [reply scripts/experiments_images/ocsvm.jl]: fixed

File: scripts/experiments_images/ocsvm.jl (hunk 4 above)
Comment (TheMrGhostman) [reply scripts/experiments_images/ocsvm.jl]: i added function to utils and reshaping will take place here

**Review by vitskvara — APPROVED** [review]

**Comment by vitskvara** [comment]
Added ocsvm, knn and abod for image data.

**Pull request merged**